File: src/main/java/org/mtransit/android/ui/PreferencesActivity.java
package org.mtransit.android.ui;

import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;

import org.mtransit.android.R;
import org.mtransit.android.commons.BundleUtils;
import org.mtransit.android.commons.LocaleUtils;
import org.mtransit.android.dev.DemoModeManager;
import org.mtransit.android.util.NightModeUtils;

import dagger.hilt.EntryPoint;
import dagger.hilt.InstallIn;
import dagger.hilt.android.AndroidEntryPoint;
import dagger.hilt.android.EntryPointAccessors;
import dagger.hilt.components.SingletonComponent;

@AndroidEntryPoint
public class PreferencesActivity extends MTActivity {

	private static final String LOG_TAG = PreferencesActivity.class.getSimpleName();

	private static final String EXTRA_SUPPORT = "extra_support";

	@NonNull
	@Override
	public String getLogTag() {
		return LOG_TAG;
	}

	private boolean showSupport = false;

	private int currentUiMode = -1;

	@NonNull
	public static Intent newInstance(@NonNull Context context) {
		return newInstance(context, false);
	}

	@NonNull
	public static Intent newInstance(@NonNull Context context, boolean support) {
		Intent intent = new Intent(context, PreferencesActivity.class);
		intent.putExtra(EXTRA_SUPPORT, support);
		return intent;
	}

	@EntryPoint
	@InstallIn(SingletonComponent.class)
	interface MTApplicationEntryPoint {
		DemoModeManager demoModeManager(); // used in attachBaseContext() before @Inject dependencies are available
	}

	@NonNull
	private MTApplicationEntryPoint getEntryPoint(@NonNull Context context) {
		return EntryPointAccessors.fromApplication(context.getApplicationContext(), MTApplicationEntryPoint.class);
	}

	@Override
	protected void attachBaseContext(@NonNull Context newBase) {
		final DemoModeManager demoModeManager = getEntryPoint(newBase).demoModeManager();
		if (demoModeManager.getEnabled()) {
			newBase = demoModeManager.fixLocale(newBase);
		} else {
			newBase = LocaleUtils.fixDefaultLocale(newBase);
		}
		super.attachBaseContext(newBase);
	}

	@Override
	protected void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		this.currentUiMode = getResources().getConfiguration().uiMode;
		setContentView(R.layout.activity_preferences);
		ActionBar actionBar = getSupportActionBar();
		if (actionBar != null) {
			actionBar.setTitle(R.string.settings);
			actionBar.setDisplayHomeAsUpEnabled(true);
		}
		restoreInstanceState(savedInstanceState, getIntent().getExtras());
	}

	private void restoreInstanceState(Bundle... bundles) {
		Boolean newShowSupport = BundleUtils.getBoolean(EXTRA_SUPPORT, bundles);
		if (newShowSupport != null) {
			this.showSupport = newShowSupport;
		}
	}

	public boolean isShowSupport() {
		return showSupport;
	}

	public void setShowSupport(boolean showSupport) {
		this.showSupport = showSupport;
	}

	@Override
	protected void onSaveInstanceState(@NonNull Bundle outState) {
		outState.putBoolean(EXTRA_SUPPORT, this.showSupport);
		super.onSaveInstanceState(outState);
	}

	@Override
	protected void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
		super.onRestoreInstanceState(savedInstanceState);
		restoreInstanceState(savedInstanceState);
	}

	@Override
	public void onConfigurationChanged(@NonNull Configuration newConfig) {
		super.onConfigurationChanged(newConfig);
		if (this.currentUiMode != newConfig.uiMode) {
			NightModeUtils.resetColorCache();
			NightModeUtils.recreate(this);
		}
	}

	@Override
	public boolean onOptionsItemSelected(@NonNull MenuItem item) {
		if (item.getItemId() == android.R.id.home) {
			finish();
			return true; // handled
		}
		return super.onOptionsItemSelected(item);
	}
}

File: src/main/java/org/mtransit/android/ui/pref/PreferencesActivity.kt
@file:JvmName("PreferencesActivity") // ANALYTICS
package org.mtransit.android.ui.pref

import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import androidx.lifecycle.lifecycleScope
import dagger.hilt.EntryPoint
import dagger.hilt.InstallIn
import dagger.hilt.android.AndroidEntryPoint
import dagger.hilt.android.EntryPointAccessors
import dagger.hilt.components.SingletonComponent
import org.mtransit.android.R
import org.mtransit.android.commons.LocaleUtils
import org.mtransit.android.dev.DemoModeManager
import org.mtransit.android.ui.MTActivity
import org.mtransit.android.util.NightModeUtils
import javax.inject.Inject

@AndroidEntryPoint
class PreferencesActivity : MTActivity(R.layout.activity_preferences) {

    companion object {
        private val LOG_TAG = PreferencesActivity::class.java.simpleName

        @JvmOverloads
        @JvmStatic
        fun newInstance(
            context: Context,
            showSupport: Boolean = false,
        ): Intent {
            return Intent(context, PreferencesActivity::class.java).apply {
                putExtra(PreferencesViewModel.EXTRA_SUPPORT, showSupport)
            }
        }
    }

    override fun getLogTag(): String = LOG_TAG

    @EntryPoint
    @InstallIn(SingletonComponent::class)
    interface PreferencesActivityEntryPoint {
        val demoModeManager: DemoModeManager // used in attachBaseContext() before @Inject dependencies are available
    }

    private fun getEntryPoint(context: Context): PreferencesActivityEntryPoint {
        return EntryPointAccessors.fromApplication(context.applicationContext, PreferencesActivityEntryPoint::class.java)
    }

    override fun attachBaseContext(newBase: Context) {
        val demoModeManager = getEntryPoint(newBase).demoModeManager
        val fixedBase = if (demoModeManager.enabled) {
            demoModeManager.fixLocale(newBase)
        } else {
            LocaleUtils.attachBaseContextActivity(newBase)
        }
        super.attachBaseContext(fixedBase)
        LocaleUtils.attachBaseContextActivityAfter(this)
    }

    @Inject
    lateinit var demoModeManager: DemoModeManager

    private var currentUiMode = -1

    override fun onCreate(savedInstanceState: Bundle?) {
        this.currentUiMode = resources.configuration.uiMode
        LocaleUtils.onCreateActivity(this)
        super.onCreate(savedInstanceState)
        supportActionBar?.apply {
            setTitle(R.string.settings)
            setDisplayHomeAsUpEnabled(true)
        }
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        if (currentUiMode != newConfig.uiMode) {
            NightModeUtils.setDefaultNightMode(context, demoModeManager)
            NightModeUtils.recreate(this)
        }
    }

    override fun onPostResume() {
        super.onPostResume()
        if (currentUiMode != resources.configuration.uiMode) {
            lifecycleScope.launchWhenResumed {
                NightModeUtils.setDefaultNightMode(context, demoModeManager)
                NightModeUtils.recreate(this@PreferencesActivity)
            }
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        return when {
            supportFragmentManager.popBackStackImmediate() -> true
            super.onSupportNavigateUp() -> true
            else -> {
                finish() // close screen
                true
            }
        }
    }
}